vtex-sites/base.store
# Feat: Store UI Theming

![Mask Group](https://user-images.githubusercontent.com/723163/155199906-d5680af4-2abf-4f1c-8613-6648367fd4c7.png)

## What's the purpose of this pull request?
This project is a **working prototype and it's not meant to be merged to the master branch yet**. It aims to validate a styling architecture that can leverage one side of the customization process—the styling—when building a new store to drastically improve the development (+ design) experience.

## How it works? 
### About customization
A store’s customization encompasses modifying/creating both features and user interfaces (function and form). Regarding the former, FastStore will provide out-of-the-box customizable components with built-in features that developers can start using right away. It’s also possible to use these basic components as a skeleton to develop more complex and tailor-made experiences. Also, a developer can create their own components for specific use cases.

As for the building and customization of the component’s user interface, FastStore has a framework-agnostic approach. All of the components come unstyled by default. So, right from the beginning, agencies can style their stores the way they want, only limited by the CSS technology (which, quite frankly, is almost unlimited).

Furthermore, it’s in our company’s best interest to help developers and partner agencies to build and customize stores rapidly and cheaply. This initiative aims to leverage these objectives by validating (then implementing) a white-label styling architecture composed of a library of standard templates for the components, alongside a comprehensive set of design tokens that, combined, form a visual theme.

Our T1 clients demand great flexibility when it comes to features and UI. By using this approach, we can offer the required adaptability in a structured manner that can help developers bootstrap a store. On top of that, this structure can be ignored, overwritten, or further extended by the developers and partner agencies.

### Base Store
Additionally, we’re developing a reference store (Base Store) to showcase the FastStore. In this effort, we’re designing and developing a UI theme, a wide range of components, and experimenting with different customization ideas and systems. Base Store is going to be a key piece on building a library of components for FastStore. The proposed project is going to be built over (and merged into) Base Store’s codebase and design system.

### Hypothesis
By offering a useful and assorted library of templates for our components and sections, alongside a comprehensive set of design tokens to further customize them, FastStore can fulfill its mission of helping developers to easily build and customize storefronts within a frictionless and productive development experience.

### Solution
This project aims to propose & validate a novel UI architecture based on templates of components and design tokens In this architecture, the store’s UI is divided into two layers: **Structure** (the components’ templates) and **Theme** (set of design tokens).

<img width="400" height="auto" align="right" src="https://user-images.githubusercontent.com/723163/155183502-8345cc09-f051-41ad-a9cc-56914b70d97a.png" />

**Structure**, the first, has foundational styles that lay things out in common design patterns for different use cases. It also covers fundamental interactions that form a component, like showing/hiding a dropdown or moving a toggle’s knob to make it active.

The second layer, **Theme**, is where the branding lives and what truly differentiates one UI from another. Graphic elements—like typography (and type scale), colors, imagery, borders, spacing, & more—are controlled by design tokens that a developer can easily tweak to create a unique look and feel.

Additional styles can be added to overwrite what was set in the Structure and/or Theme, reaching a state of total flexibility and freedom. Lastly, the developer can throw away all existing FastStore CSS altogether and build it from scratch.

## How to test it?
* Access [/theming-poc](http://localhost:8000/theming-poc) to see the first components that make use of the design tokens + theming structure;
* Access [/theming-poc-homepage](http://localhost:8000/theming-poc-homepage) to see a simpler, themed version of the homepage, and insert one of the following theme classes to the page's `body`:
  * `.new-theme`: used for validation purpose;
  * `.base-store`: Base Store's aesthetics;
  * `.grocery`: Grocery Starter Pack's aesthetics;
  * `.nucommerce`: another theme to assess the flexibility of the overall theming structure.
*  Check the theme files to have a grasp of global design tokens:
   * `styles/themes/new-theme.scss`
   * `styles/themes/base-store.scss`
   * `styles/themes/grocery.scss`
   * `styles/themes/nucommerce.scss`
